PEP 740: Feedback, round 2 (#3692)
python/peps: peps/pep-0740.rst

  Digital attestations impose additional sophistication requirements: the
  attacker must be sufficiently sophisticated to access private signing material
  (or signing identities).

> * Index verifiability: in the status quo, the only attestation provided by the

Author: Sounds good, thanks! I'll add a sentence making that explicit.

  index is an optional PGP signature per release file
  (see :ref:`PGP signatures <pgp-signatures>`). These signatures are not
  checked by the index either for well-formedness or for validity, since
  the index has no mechanism for identifying the right public key for the
  signature.

This PEP recommends a generic attestation format, including a serializable

> payload format for signature generation, with the expectation that index providers adopt the

Author: Yeah, good call -- I'll also try and unify the language here with what we use elsewhere.

  the index has no mechanism for identifying the right public key for the
  signature.

This PEP recommends a generic attestation format, including a serializable

> payload format for signature generation, with the expectation that adopt the

Reviewer: This sentence is missing a subject. I don't understand the topic well enough to know what is meant, but maybe:
```suggestion
payload format for signature generation, with the expectation that index providers adopt the
```

Author: Thanks! Yeah, that's the intended subject.

In practice, this means that ``distribution`` is defined by the PyPA's
living specifications for
:ref:`binary distributions <packaging:binary-distribution-format>` and
:ref:`source distributions <packaging:source-distribution-format>`, although
non-conforming distributions may be hosted by the index.

The following pseudocode demonstrates the construction of an attestation
payload and its signature:

.. code-block:: python

  def build_payload(dist: Path) -> AttestationPayload:
      return AttestationPayload(
        distribution=dist.name,
        digest=sha256(dist.read_bytes()).hexdigest,
      )

  attestation_payload = build_payload("sampleproject-1.2.0-py2.py3-none-any.whl")

  # canonical_json is a fictitious module that performs RFC 8785 canonical
  # JSON serialization.
  encoded_payload = canonical_json.dumps(asdict(attestation_payload))

  raw_signature = signing_key.sign(encoded_payload, ECDSA(SHA2_256()))
  message_signature = b64encode(raw_signature)

.. _provenance-object:

Provenance objects
------------------

The index will serve uploaded attestations along with metadata that can assist
in verifying them in the form of JSON serialized objects.

> These *provenance objects* will be available via both the :pep:`503` Simple Index

Author: Yep, this got shifted around. It should be above 🙂

* When an uploaded file has one or more attestations, the index **MAY** include a
  ``provenance`` object in the ``file`` dictionary for that file.
* ``provenance``, when present, **MUST** be a :ref:`provenance object <provenance-object>`.

These changes require a version change to the JSON API:

* The ``api-version`` must specify version 1.2 or later.

.. _attestation-object:

Attestation objects
-------------------

An attestation object is a JSON object with several required keys; applications
or signers may include additional keys so long as all explicitly
listed keys are provided. The required layout of an attestation
object is provided as pseudocode below.

.. code-block:: python

  @dataclass
  class Attestation:
      version: Literal[1]
      """
      The attestation format's version, which is always 1.
      """

      verification_material: VerificationMaterial
      """
      Cryptographic materials used to verify `message_signature`.
      """

      message_signature: str
      """
      The attestation's signature, as `base64(raw-sig)`, where `raw-sig`
      is the raw bytes of the signing operation.
      """

  @dataclass
  class VerificationMaterial:

>       certificate: str

Reviewer 2: Do we want to allow for keys, or keep this focused on identity-based signing?

Author: I was going to limit this to identity based signing for the "v1" format, just in terms of reducing the scope of the PEP (so that I don't have to specify PyPI's handling of keys, since that doesn't currently exist 😅).

Author: I thought about this some more, and unless there's a strong objection to the current identity cert limitation, IMO it's the most tractable approach for the "v1" object format (in terms of keeping the PEP relatively small + reflecting what the initial PyPI implementation will do).

Reviewer 2: SGTM!

      """
      The signing certificate, as `base64(DER(cert))`.
      """

      transparency_entries: list[TransparencyLogEntry]
      """

>       Zero or more transparency log entries for this attestation's signature

Reviewer 2: I would have this be one or more, and say that zero is allowed only when your threat model shows that signers and consumers are in the same trust boundary.

Author: Sounds good to me!

Author: Done!

      and certificate.

>       """

Reviewer 2: I would also specify one or more signed timestamps to verify the certificate's validity (where signed timestamps can take multiple forms, RFC3161, Roughtime, a log entry, etc)

Author: Makes sense. I think specifying the shape of each of these might get tricky, however, especially for the ones that don't have a standard JSON layout 🙂 

Instead of specifying each, what do you think about language (roughly) like this?

> Each attestation object MUST come with one or more independent, signed sources of time for the attestation. Examples of this signed timestamp include transparency log entries, RFC 3161 timestamps, and Roughtime signatures. Verifiers SHOULD verify these timestamp sources as part of establishing the attestation's timeliness, but MAY choose not to for support or policy reasons.

That gives us flexibility in the "v1" attestation here, which we can then ratchet down to a "MUST" in a subsequent version (along with a better-defined set of the sources/layouts of signed time).

WDYT?

Reviewer 2: I like it, the flexibility sounds good.
I'm also happy with stating "one" and not "zero" in this draft. I know we talked about the option to say zero if you understand the threat model, I think that can be left for a future revision that is focused on BYO PKI.

      """
      The kind (type) of log entry.
      """

      entry_version: str
      """
      The version of the log entry's submitted format.
      """

      integrated_time: int
      """
      The UNIX timestamp from the log from when the entry was persisted.
      """

      inclusion_proof: InclusionProof
      """
      The actual inclusion proof the the log entry.
      """

  @dataclass
  class InclusionProof:
      log_index: int
      """
      The index of the entry in the tree it was written to.
      """

      root_hash: str
      """
      The digest stored at the root of the Merkle tree at the time of proof
      generation.
      """

      tree_size: int
      """
      The size of the Merkle tree at the time of proof generation.
      """

      hashes: list[str]
      """
      A list of hashes required to complete the inclusion proof, sorted
      in order from leaf to root. The leaf and root hashes are not themselves
      included in this list; the root is supplied via `root_hash` and the client
      must calculate the leaf hash.
      """

      checkpoint: str
      """
      The signed tree head's signature, at the time of proof generation.

>       """

Author: Flagging: I'm not sure if there's value in fully specifying these, or instead defining them behaviorally + leaving them as opaque objects in the PEP. Their structure is arguably an implementation detail, one with potential to change as logs (and log implementations) change.

Reviewer 2: The only other field for completeness would be a list of cosigned checkpoints from witnesses. There is not a published spec yet for the format.

Author: Hmm, any thoughts on how I should represent that? `cosigned_checkpoints: list[object]`?

Reviewer 2: `list[str]` should be good, it's going to look similar to the checkpoint format.

Attestation objects are versioned; this PEP specifies version 1. Each version
is tied to a single cryptographic suite to minimize unnecessary cryptographic
agility. In version 1, the suite is as follows:

* Certificates are specified as X.509 certificates, and comply with the
  profile in :rfc:`5280`.
* The message signature algorithm is ECDSA, with the P-256 curve for public keys
  and SHA-256 as the cryptographic digest function.

Future PEPs may change this suite (and the overall shape of the attestation
object) by selecting a new version number.

Attestation payload and signature generation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The *attestation payload* is the actual claim that is cryptographically signed
over within the attestation object (as the ``message_signature``).

The attestation payload is encoded as an :rfc:`8785` canonicalized JSON object,

> with the following pseudocode layout:

Author: Flagging: technically the only property we need here is canonicalizability. The PEP currently specifies RFC 8785 since it's a compatible representation of JSON, but I'm not sure if there's a mature implementation of JCS/8785 for Python.

Alternatives here would be DER (since we already have X.509 in the mix) or something bespoke here (e.g. a plain byte string with `:` or `\n` domain separators between fields).

Author: Just to dump my research here, here are the JCS implementations I found for Python:

* https://github.com/titusz/jcs (available on PyPI, last updated 2 years ago)
* https://github.com/cyberphone/json-canonicalization/tree/master contains a Python implementation, but not published to PyPI
* https://github.com/AnasMK/json-canonical forked from the above, targets Python 2?
* https://github.com/sveinugu/jsoncanon on PyPI, appears actively maintained, doesn't have FP support (which we probably don't need)

Of these, `jcs` and `jsoncanon` are probably the most immediately usable, but both are relatively small projects.

Author: There's also `encode_canonical` from securesystemslib, but AFAICT this doesn't implement JCS -- it implements a different but similarly named canonicalization scheme from OLPC: <https://wiki.laptop.org/go/Canonical_JSON>.

Author: Update: I'm working on a minimal implementation of RFC 8785 that should be a maintainable, stable basis for this PEP.

Author: I've created https://github.com/trailofbits/rfc8785.py, which should be sufficient for our needs. So I think this is no longer an issue.

  the index **SHALL** accept ``attestations`` as an additional multipart form
  field.
* The new ``attestations`` field **SHALL** be a JSON array.
* The ``attestations`` array **SHALL** have one or more items, each a JSON object
  representing an individual attestation.
* Each attestation object **MUST** be verifiable by the index. If the index fails
  to verify any attestation in ``attestations``, it **MUST** reject the upload.
  The format of attestation objects is defined under :ref:`attestation-object`
  and the process for verifying attestations is defined under
  :ref:`attestation-verification`.

Index changes
-------------

Simple Index
^^^^^^^^^^^^

* When an uploaded file has one or more attestations, the index **MAY** include a
  ``data-provenance`` attribute on its file link.

* When the ``data-provenance`` attribute is present, its value **SHALL** be
  a base64 encoded :ref:`provenance object <provenance-object>`.

JSON-based Simple API
^^^^^^^^^^^^^^^^^^^^^

* When an uploaded file has one or more attestations, the index **MAY** include a
  ``provenance`` object in the ``file`` dictionary for that file.

> * ``provenance``, when present, **MUST** be a :ref:`provenance object <provenance-object>`.

Reviewer 3: Might want to make it more clear here that only a single attestation can be considered the provenance object.

Author: I intended the opposite here 😅 -- a provenance object is meant to be `(identity, attestations)`, i.e. the Trusted Publisher identity that uploaded the file and one or more attestations that came from that identity (or others, such as a 3p attestation or countersignature).

Reviewer 3: I think this could be made more clear then, this reads as if the index is exposing a single `provenance` object.

Author: Done -- I've emphasized that a provenance object is composed of attestations, and that the each file exposes its own provenance object.

            },
            [
              { /* attestation 1 ... */ },
              { /* attestation 2 ... */ }
            ]
          ]
        ]
    }

or, as pseudocode:

.. code-block:: python

  @dataclass
  class Publisher:
      kind: string
      """
      The kind of Trusted Publisher.
      """

      claims: object | None
      """
      Any context-specific claims retained by the index during Trusted Publisher
      authentication.
      """

      _rest: object
      """
      Each publisher object is open-ended, meaning that it MAY contain additional
      fields beyond the ones specified explicitly above. This field signals that,
      but is not itself present.
      """

  @dataclass
  class Provenance:
      version: Literal[1]
      """
      The provenance object's version, which is always 1.
      """

      attestations: list[tuple[Publisher, list[Attestation]]]
      """
      One or more attestation objects, indexed by their publisher.
      """

* ``version`` is ``1``. Like attestation objects, provenance objects are
  versioned, and this PEP only defines version ``1``.
* ``attestations`` is a **required** JSON array, containing pairs

>   of publishers and a JSON array of one or more attestation objects.

Reviewer 4: Why do we use a tuple of `(publisher, List[attestations])`, and not an actual object like:
```JSON
{
    "publisher": {....}
    "attestations": [...]
}
```
(I'm mostly thinking that the object would be easier to extend in the future, rather than a tuple where the meaning is determined by the element order)

Author: No good reason -- your suggestion is much cleaner 🙂.

How does this sound?

```
class PublisherAttestations:
  publisher: Publisher
  attestations: list[Attestation]

class Provenance:
  version: Literal[1]
  attestations: list[PublisherAttestations]
```

(The `PublisherAttestations` name might need workshopping.)

Reviewer 4: Yeah. I also think we should rename the `attestations` field in `Provenance`, since it's not _only_ providing attestations. And with the current naming, you'd have things like:
```python
provenance_object['attestations'][0]['attestations'][0]
```
due to `attestations` being the name of the field in both structs.

Since what we want to encode is that each Attestation object is associated with some metadata needed to verify it (in this case, a `Publisher`), we could have something like:

```
class AttestationBundle:
  publisher: Publisher
  attestations: list[Attestation]

class Provenance:
  version: Literal[1]
  attestation_bundles: list[AttestationBundle]
```

In this case, `AttestationBundle` is a bundle of `Attestations` + some metadata related to them. But yeah, I'm also open to better naming suggestions.

Author: SGTM. `attestation_bundles` is much better than overloading `attestations`, and we can iterate if a better name comes along 🙂

  Each publisher is a Trusted Publishing identity, as described
  by the ``Publisher`` model above. As noted in the ``Publisher`` model,
  each publisher object is specific to its Trusted Publisher but must include
  at minimum:

  * A ``kind`` key, which **MUST** be a JSON string that uniquely identifies the
    kind of Trusted Publisher.
  * A ``claims`` key, which **MUST** be a JSON object containing any context-specific
    claims retained by the index during Trusted Publisher authentication.

  All other keys in the publisher object are publisher-specific. A full
  illustrative example of a publisher object is provided in :ref:`appendix-1`.

  Each array of attestation objects is a superset of the ``attestations``
  array supplied by the uploaded through the ``attestations`` field at upload
  time, as described in :ref:`upload-endpoint` and
  :ref:`changes-to-provenance-objects`.

.. _changes-to-provenance-objects:

Changes to provenance objects
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Provenance objects are *not* immutable, and may change over time. Reasons

> for changes to the provenance object include but are not limited to:

Reviewer 4: The examples mentioned below of the provenance objects changing are only about _adding_ new things (attestations, publishers). Does that mean that provenance objects are mutable, but append-only? Or can attestations be _removed_ too?

Author: As specified, they're append-only. I can imagine scenarios in which the index _might_ want to remove attestations however, so I didn't want to formally guarantee that the index never does so.

* Addition of new attestations: the index **MAY** choose to support attestations
  from sources other than the file's uploader, such as attestations by
  third-party auditors or the index itself. These attestations may be performed
  asynchronously or only appear after the file is already public on the index,
  requiring the index to insert them into the provenance *post facto*.

* Addition of signing identities: when new attestations are added to the
  provenance object, their verification may require new signing identities

>   (such as public keys or Trusted Publisher identities).

Author: This is badly written: what I was thinking of here is index signatures (which might use a pubkey, or an OIDC identity), or 3p signatures (e.g. reproducers, auditors), which also might be either pubkeys or OIDC. In either of those cases, we'll need to add the entity's signing identity (pubkey or OIDC) to the provenance object.

(I'll reword here.)

``verification_material.transparency_entries`` on a policy basis, e.g. requiring
at least one transparency log entry or a threshold of entries. When verifying
transparency entries, the verifier **MUST** confirm that the inclusion time for
each entry lies within the signing certificate's validity period.

Security Implications
=====================

This PEP is primarily "mechanical" in nature; it provides layouts for
structuring and serving verifiable digital attestations without specifying
higher level security "policies" around attestation validity, thresholds
between attestations, and so forth.

Cryptographic agility in attestations
-------------------------------------

Algorithmic agility is a common source of exploitable vulnerabilities
in cryptographic schemes. This PEP limits algorithmic agility in two ways:

* All algorithms are specified in a single suite, rather than a geometric
  collection of parameters. This makes it impossible (for example) for an
  attacker to select a strong signature algorithm with a weak hash function,
  compromising the scheme as a whole.
* Attestation objects are versioned, and may only contain the algorithmic

>   suite or suites specified for their version. If a specific suite

Reviewer 4: Above, in the `Attestation objects` section, we say:
> Each version is tied to a **single** cryptographic suite to minimize unnecessary cryptographic agility.

So here we should remove the `or suites`,  or change the above to say that a version can be tied to multiple suites.
```suggestion
  suite specified for their version. If a specific suite
```

Author: Yes, thanks!

These *provenance objects* will be available via both the :pep:`503` Simple Index
and :pep:`691` JSON-based Simple API as described above, and will have the
following layout:

.. code-block:: json

    {
        "version": 1,
        "attestation_bundles": [
          {
            "publisher": {
              "kind": "important-ci-service",
              "claims": {},
              "vendor-property": "foo",
              "another-property": 123
            },
            "attestations": [
              { /* attestation 1 ... */ },
              { /* attestation 2 ... */ }
            ]
          }
        ]
    }

or, as pseudocode:

.. code-block:: python

  @dataclass
  class Publisher:
      kind: string
      """
      The kind of Trusted Publisher.
      """

      claims: object | None
      """
      Any context-specific claims retained by the index during Trusted Publisher
      authentication.
      """

      _rest: object
      """
      Each publisher object is open-ended, meaning that it MAY contain additional
      fields beyond the ones specified explicitly above. This field signals that,
      but is not itself present.
      """

  @dataclass

>   class AttestationBundle:
>       publisher: Publisher
> 
>       attestations: list[Attestation]
> 

Reviewer 4: ```suggestion
  class AttestationBundle:
      publisher: Publisher
      """
      The publisher associated with this set of attestations.
      """

      attestations: list[Attestation]
      """
      The set of attestations included in this bundle.
      """

```

  @dataclass
  class Provenance:
      version: Literal[1]
      """
      The provenance object's version, which is always 1.
      """

      attestation_bundles: list[AttestationBundle]
      """
      One or more attestation "bundles".
      """

* ``version`` is ``1``. Like attestation objects, provenance objects are
  versioned, and this PEP only defines version ``1``.

> * ``attestation_bundles`` is a **required** JSON array, containing one
>   or more "bundles" of attestations, paired with their signing identity
>   (such as a Trusted Publishing identity).

Reviewer 4: Another possible wording:

```suggestion
* ``attestation_bundles`` is a **required** JSON array, containing one
  or more "bundles" of attestations. Each bundle corresponds to a
  signing identity (such as a Trusted Publishing identity), and contains
  one or more attestation objects.
```

To try to make clear that besides having multiple bundles, each bundle can have multiple attestations and the "bundle" is a way of grouping them by signing identity

  to verify any attestation in ``attestations``, it **MUST** reject the upload.
  The format of attestation objects is defined under :ref:`attestation-object`
  and the process for verifying attestations is defined under
  :ref:`attestation-verification`.

Index changes
-------------

Simple Index
^^^^^^^^^^^^

* When an uploaded file has one or more attestations, the index **MAY**
  provide a ``.provenance`` file adjacent to the hosted distribution.
  The format of the ``.provenance`` file **SHALL** be a JSON-encoded
  :ref:`provenance object <provenance-object>`, which **SHALL** contain
  the file's attestations.

  For example, if an uploaded file is hosted at
  the URL ``https://example.com/sampleproject-1.2.3.tar.gz``, the provenance
  URL would be ``https://example.com/sampleproject-1.2.3.tar.gz.provenance``.

* When a ``.provenance`` file is present, the index **MAY** include a
  ``data-provenance`` attribute on its file link. The value of the
  ``data-provenance`` attribute **SHALL** be the SHA256 digest of the
  associated ``.provenance`` file.

* The index **MAY** choose to modify the ``.provenance`` file. For example,

>   the index **MAY** choose to add additional attestations and verification

Reviewer 3: ```suggestion
  the index **MAY** permit adding additional attestations and verification
```

  All other keys in the ``publisher`` object are publisher-specific. A full
  illustrative example of a ``publisher`` object is provided in :ref:`appendix-2`.
* ``attestations`` is a **required** JSON array, containing one or
  more attestation objects. This array is a superset of ``attestations`` array
  supplied by the uploader through the ``attestations`` field at upload time,
  as described in :ref:`upload-endpoint`.

  Because ``attestations`` is a superset of the file's original uploaded attestations,
  the index **MAY** chose to embed additional attestations. These attestations
  may come from the index itself, from third parties, or as subsequent actions
  produced by the original uploader.

Simple Index
^^^^^^^^^^^^

* When an uploaded file has one or more attestations, the index **MAY** include a
  ``data-provenance`` attribute on its file link.

* When the ``data-provenance`` attribute is present, its value **SHALL** be

>   a base64 encoded :ref:`provenance-object`.

Author: CC @di: this will embed the provenance object directly into the PEP 503 simple index as you suggested, but I want to call out that it'll likely be quite large there 🙂. I'll do some research to see how well/gracefully typical HTML5 parsers handle large data attributes.

          "vendor-property": "foo",
          "another-property": 123
        },
        "attestations": [
          { /* ... */ },
          { /* ... */ }
        ]
    }

or, as pseudocode:

.. code-block:: python

  @dataclass
  class Provenance:
      version: Literal[1]
      """
      The provenance object's version, which is always 1.

>       """

Author: NB: This adds a `version` to the provenance object as well. Per @haydentherapper's note, this makes future extensibility with self-held keys easier (since these would presumably fall under their own fields, not under `publisher`).

  produced by the original uploader.

.. _attestation-verification:

Attestation verification
------------------------

Verifying an attestation object requires verification of each of the following:

* ``verification_material.certificate`` is a valid signing certificate, as
  issued by an *a priori* trusted authority (such as a root of trust already
  present within the verifying client).
* ``verification_material.certificate`` identifies an appropriate signing
  subject, such as the machine identity of the Trusted Publisher that published
  the package.
* ``message_signature`` can be verified by ``verification_material.certificate``,
  using the reconstructed attestation payload as the cleartext input. The
  verifier **MUST** reconstruct the attestation payload itself.

In addition to the above required steps, a verifier **MAY** additionally verify
``verification_material.transparency_entries`` on a policy basis, e.g. requiring
at least one transparency log entry or a threshold of entries. When verifying
transparency entries, the verifier **MUST** confirm that the inclusion time for
each entry lies within the signing certificate's validity period.

> 

Reviewer 2: Do we need a description of a trusted root file?

Author: I *think* we don't -- the reasoning I'm applying here is that the trust root is essentially an implementation detail in the layer of "you need an a priori source of trust," similar to how CABF doesn't specify anything about OS roots other than that they have the basic trust anchor contents needed to construct chains.
